### PR TITLE
Tour translation

### PR DESCRIPTION
**Pull request by sam33rdhakal**
Updates:
* Changes the tour details when the tour language is changed
* Saves/restores user selected tour language when a tour is started/resumed
* Prioritizes tour locale over app locale on tour related screens

### CODE REVIEW

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt [hunk 1]
```diff
@@ -44,13 +44,16 @@ class LanguageSelector(private val prefs: LocalizationPreferences) {
...
-
         // TODO: Investigate possibility of replacing the list with a `LocaleList`
         return languages.firstOrNull {
             // This is very much intentional. Read the method docs fully before changing.
-            it.underlyingLocale().language == appLocale.language
+            if (prioritizeTour) {
+                it.underlyingLocale().language == prefs.tourLocale.language
+            } else {
+                it.underlyingLocale().language == appLocale.language
+            }
```
Comment (Cliabhach) [inline localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: So, this doesn't quite work. We'll need two loops in some circumstances.

Imagine we are looking at the audio details for one of the stops in a tour. We are working under the environment immediately following:

Available Languages for the audio:
* English
* Chinese

Tour Language selection:
* Spanish

App Language selection:
* Chinese

With the code as given, we will loop through once, not find Spanish, and fall back directly to English. With two loops, we could then search for Chinese and find it successfully.

File: tours/src/main/kotlin/edu/artic/tours/TourDetailsViewModel.kt [hunk 2]
```diff
@@ -136,8 +130,11 @@ class TourDetailsViewModel @Inject constructor(
      * Navigate user to the Map activity in Tour Context.
      */
     fun onClickStartTour() {
-        tourObservable.take(1)
-                .subscribe { tour ->
+        Observables.combineLatest(tourObservable, chosenTranslation.map { it.underlyingLocale() })
+                .take(1)
+                .subscribe { tourWithLocale ->
```
Comment (Cliabhach) [inline tours/src/main/kotlin/edu/artic/tours/TourDetailsViewModel.kt]: You could decompose here with `.subscribeBy { (tour, locale) ->` instead. What do you think?

File: db/src/main/kotlin/edu/artic/db/models/ArticTour.kt [hunk 3]
```diff
@@ -131,6 +132,29 @@ data class ArticTour(
         return this.title
     }
 
+    /**
+     * Returns the translations including english language
+     */
+    val translationWithEnglish: List<Translation>
```
Comment (Cliabhach) [inline db/src/main/kotlin/edu/artic/db/models/ArticTour.kt]: Please rename to `allTranslations` to match the naming used by `ArticAudioFile`.

File: db/src/main/kotlin/edu/artic/db/models/ArticTour.kt [hunk 4]
```diff
@@ -131,6 +132,29 @@ data class ArticTour(
...
+        get() {
+            return mutableListOf(this.toEnglishTranslation()).apply { addAll(translations) }
+        }
+
+}
+
+/**
+ * Builds english translation for [ArticTour]
```
Comment (Cliabhach) [inline db/src/main/kotlin/edu/artic/db/models/ArticTour.kt]: Edit comment to make it clear that we are returning, essentially, a copy of `this`.

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt [hunk 5]
```diff
@@ -26,6 +26,15 @@ class LanguageSelector(private val prefs: LocalizationPreferences) {
         }
     }
 
+    fun setTourLanguage(lang: Locale) {
```
Comment (Cliabhach) [inline localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: Rename `lang` to something that makes the `hasNoLanguage()` call read better. Maybe `proposed`?

File: localization/src/main/kotlin/edu/artic/localization/LocalizationPreferences.kt [hunk 6]
```diff
@@ -14,6 +14,10 @@ import java.util.*
 class LocalizationPreferences(context: Context)
     : BasePreferencesManager(context, "localization") {
 
+    companion object {
+        const val TOUR_KEY = "tour_locale"
```
Comment (Cliabhach) [inline localization/src/main/kotlin/edu/artic/localization/LocalizationPreferences.kt]: The name looks an awful lot like `ARG_TOUR`. Do you think the fact that this is in `LocalizationPreferences.kt` will be enough to distinguish the two?

File: map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt [hunk 7]
```diff
@@ -24,7 +25,8 @@ import javax.inject.Inject
 class TourCarouselViewModel @Inject constructor(private val analyticsTracker: AnalyticsTracker,
                                                 private val objectDao: ArticObjectDao,
                                                 private val audioObjectDao: ArticAudioFileDao,
-                                                val tourProgressManager: TourProgressManager) : BaseViewModel() {
+                                                val tourProgressManager: TourProgressManager,
+                                                private val languageSelector: LanguageSelector) : BaseViewModel() {
```
Comment (Cliabhach) [inline map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt]: Optional: put the `languageSelector` on the line above, with the other `private val`s. It'll make the git diff cleaner.

File: map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt [hunk 8]
```diff
@@ -47,7 +49,14 @@ class TourCarouselViewModel @Inject constructor(private val analyticsTracker: An
         }
 
     init {
-        tourObservable.map { it.title }
+        tourObservable
+                .map{tour->tour.translationWithEnglish}
```
Comment (Cliabhach) [inline map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt]: Indentation?

File: tours/src/main/kotlin/edu/artic/tours/TourDetailsViewModel.kt (hunk 2 above)
Comment (sam33rdhakal) [reply tours/src/main/kotlin/edu/artic/tours/TourDetailsViewModel.kt]: Sure. Looks cleaner. Thanks

File: db/src/main/kotlin/edu/artic/db/models/ArticTour.kt (hunk 3 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/models/ArticTour.kt]: Refactored.

File: db/src/main/kotlin/edu/artic/db/models/ArticTour.kt (hunk 4 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/models/ArticTour.kt]: Edited.

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt (hunk 5 above)
Comment (sam33rdhakal) [reply localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: Renamed it to `setTourLanguage`.

File: localization/src/main/kotlin/edu/artic/localization/LocalizationPreferences.kt (hunk 6 above)
Comment (sam33rdhakal) [reply localization/src/main/kotlin/edu/artic/localization/LocalizationPreferences.kt]: That name is awful 😄.  Updated it to PREF_TOUR_LOCALE. You are absolutely right sir.

File: map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt (hunk 8 above)
Comment (sam33rdhakal) [reply map/src/main/kotlin/edu/artic/map/carousel/TourCarouselViewModel.kt]: Fixed.

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt [hunk 9]
```diff
@@ -26,12 +26,12 @@ class LanguageSelector(private val prefs: LocalizationPreferences) {
...
-    fun setTourLanguage(lang: Locale) {
-        if (lang.hasNoLanguage()) {
+    fun setTourLanguage(proposedTourLocale: Locale) {
+        if (proposedTourLocale.hasNoLanguage()) {
             prefs.remove(LocalizationPreferences.TOUR_KEY)
         } else {
-            prefs.tourLocale = lang
-            appLocaleRef.set(lang)
+            prefs.tourLocale = proposedTourLocale
+            appLocaleRef.set(proposedTourLocale)
```
Comment (Cliabhach) [inline localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: Ah, we shouldn't change `appLocaleRef` here....

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt (hunk 1 above)
Comment (sam33rdhakal) [reply localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: Thank you for detailed explanation @Cliabhach 👍 
Updated it. Please review it.

File: tours/src/main/res/layout/fragment_tour_details.xml [hunk 10]
```diff
@@ -33,7 +33,7 @@
                 android:layout_height="wrap_content"
                 android:layout_marginTop="@dimen/marginStandard"
                 android:layout_marginEnd="@dimen/marginStandard"
-                android:background="@null"
+                android:background="@color/audioBackground"
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply tours/src/main/res/layout/fragment_tour_details.xml]: There are a bunch of colors which are named in a similar manner. We could work on these later.

File: localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt [hunk 11]
```diff
@@ -35,14 +43,25 @@ class LanguageSelector(private val prefs: LocalizationPreferences) {
...
-
         // TODO: Investigate possibility of replacing the list with a `LocaleList`
+
+        val priority: T? = if (prioritizeTour) {
+            findIn(languages, prefs.tourLocale)
+        } else {
+            null
+        }
+
+        return priority ?: findIn(languages, appLocale) ?: languages.first()
```
[the remark this answers is not in the document]
Comment (sam33rdhakal) [reply localization/src/main/kotlin/edu/artic/localization/LanguageSelector.kt]: Thanks 😄

**Pull request merged**